Readme.txt
ZDZ 的程序的简化版的说明, v2.0.13
最后更新 2012-05-22, xyy

使用 raster_tuning --help, 查看用的是普通 LIF 模型还是指数LIF模型(开头会有一句 "! Use Exponential I&F model")

程序改成只有边显示图形边计算及纯计算两个模式. 其他如计算 Lyapunov 指数,
 功率谱等功能皆已去除(但愿去除干净了), 即是说关于分析的代码去除了.

程序默认会把完整的电平数据输出到 data 目录下的 staffsave.txt. 每次运行会覆盖上次数据
 如果需要输出放电时刻的数据, 需在执行时加上参数 --save-spike FILE
 如需要把电平数据输出到别处, 需用参数 -o FILE 指定路径
 参见下面 "程序命令行选项" 一节

默认配置用 --help 参数查看.

程序命令行选项:
  语法:
  raster_tuning [选项]
    -ng           不显示图形界面
    -t TIME       总计算时间, TIME 以毫秒为单位. (默认值为原来的宏COMP_TIME)
    -n N [N2]     神经元数, N 是兴奋型, N2 是抑制型. 省略 N2 相当于 N2=0
    -inf FILE     设置配置文件的路径. 默认是 test2.txt
                  特别地, -inf - 表示不读取任何文件, 使用程序内部的默认值.
    -mat FILE     设置连接矩阵的路径. 默认是 cortical_matrix.txt
                  特别地, -mat - 表示路径是完全图
    -o FILE       设置输出电平文件的路径, 若路径中的目录不存在, 程序会自动建立.
                  默认是 data/staffsave.txt
    --save-conductance FILE
                  输出电导(兴奋电导)数据到文件 FILE
    --save-spike FILE
                  设置输出放电时刻的文件路径. 默认不输出. FILE 为 "-" (短横杠) 时,
                  表示输出到 "./data/ras.txt"
    --save-spike-interval FILE
                  保存平均放电间隔到文件 FILE
    --read-pr FILE
                  从文件 FILE 读取每个神经元的 poisson 输入率(相对值).
    --read-ps FILE
                  从文件 FILE 读取每个神经元的 poisson 输入强度(相对值).
    -s, --save-while-cal
                  设置边计算边输出电平数据(默认)
    --save-last   设置在最后输出电平数据, 输出数据不完整. 仅用作测试.
    --bin-save    电压数据保存成二进制格式(raw double)
    -scee VALUE   设置兴奋型到兴奋型的连接强度
    -scei VALUE   设置兴奋型到抑制型的连接强度
    -scie VALUE   设置抑制型到兴奋型的连接强度
    -scii VALUE   设置抑制型到抑制型的连接强度
    -pr VALUE     设置 poisson 型输入的频率
    -ps VALUE     设置 poisson 型输入对兴奋型神经的单次强度
    -psi VALUE    设置 poisson 型输入对抑制型神经的单次强度
    --pr-mul [VALUE ...] [VALUE@POSITION ...]
                  在命令行设置各个神经元的 poisson 输入率倍率. 可以只设定前几个.
                  可以用 VALUE@POSITION 的语法指定第几个的值, 编号从 1 开始
                  未指定值的神经默认是 1. 必须提前指定神经元个数(即通过 -n 选项).
                  例如: --pr-mul 1.1 1.2 1.3@4 1.4@2 1.5 1.6 的结果相当于
                       --pr-mul 1.1 1.4 1.5 1.6 (即是说 "1.2 1.3@4" 被后面的值覆盖了)
    --ps-mul [VALUE ...] [VALUE@POSITION ...]
                  在命令行设置各个神经元的 poisson 输入强度(兴奋型)倍率. 语法同 --pr-mul
    --psi-mul [VALUE ...] [VALUE@POSITION ...]
                  在命令行设置各个神经元的 poisson 输入强度(抑制型)倍率. 语法同 --pr-mul
    -dt VALUE     设置时间步长为 VALUE
    --save-interval VALUE 或 -stv VALUE
                  输出数据的记录间隔. 若使用 --RC-filter, dt 会被更改为最近整数分之一.
    --seed-auto-on, --seed-auto-off
                  打开或关闭自动设置随机种子(用微秒级(windows上是毫秒级)的系统时间). 默认是打开
    -seed VALUE   设置随机数种子, 隐含--seed-auto-off. 功能同配置文件中的 initial_seed
                  当 seed 设为 0 时, 相当于使用 test2.txt 中的 initial_seed.
    --RC-filter   使用 RC 低通滤波后再采样. 截止频率为"输出数据的记录间隔"对应的最高频率
                  使用此选项时, dt 会强制设成 stv 的整数分之一.
    --RC-filter VALUEco VALUEci
                  使用 RC 低通滤波, 手动设置滤波系数. y[t] = co * y[t-dt] + ci * x[t]
                  输出的数据是 y[stv], y[2*stv], ..., y[k*stv]
                  其中 stv 是输出数据的记录间隔. 见 --save-interval
    -v, --verbose 计算时输出更多信息
    -q, --quiet   安静模式, 只显示错误和警告
    -h, --help    显示帮助(以及一些默认值), 然后程序终止.
    --version     显示版本信息, 然后程序终止.

程序读取输入参数的顺序是
  1. 命令行
  2. 配置文件(默认是 test2.txt)
  3. 程序内置的默认值
以先读到的值为准. 命令行中若出现重复或互斥的项, 以最后(右侧)的为准.

--pr-mul 与 --read-pr 的作用是累积的.

命令行参数以及输入输出的文件中, 神经元编号从 1 开始. (虽然程序内部是从 0 开始的)

注意:
    2011-10-22 以前的版本, 放电时刻文件神经元编号从 0 开始. 之后的从 1 开始.
    输出的数据开头会有一段未平稳的状态, 未去除.
    从 2.0.13 开始, 当使用参数 --RC-filter 时, 输出的电压数据会根据精确的放电时刻进行平均, 从而更为平滑.
    参数文件读取对于不符合格式(malformat)的文件可能会出问题. -inf, -mat, --read-pr, --read-ps
    图形模式时显示的 Poisson 输入频率及输入强度是指第一个神经元的. 但用键盘更改时所有值同比例变化.
    抑制型的神经尚未完整测试

示例:
raster_tuning -ng -n 20 -t 10000 -inf test2_0.txt -o data/staffsave2.txt -v
 含义是不显示图形界面, 20个神经元(都是兴奋型), 计算到 10000ms,
     配置(参数)文件是 test2_0.txt, 输出到 data/staffsave2.txt, 显示更多信息.

raster_tuning -n 2 -pr 1 -ps 0.012 --ps-mul 0.1@2
 显示图形界面, 第二个神经的Poisson输入强度设为 0.1 * 0.012 (第一个仍是 0.012)

raster_tuning -h
 显示帮助然后退出


默认配置文件:
test.txt            周期电流输入, 宏POISSON_INPUT_USE 为 0 时启用
test2.txt           Poisson 脉冲输入, 宏POISSON_INPUT_USE 为 1 时启用
cortical_matrix.txt 神经元间连接关系矩阵, 宏CORTICAL_STRENGTH_NONHOMO 为 1 时启用


神经连接临接矩阵文件 note:
  1. 第 i 行 j 列的元素表示 j 神经元对 i 神经元的影响. 元素的值不必是整数.
  2. 对角线的值不会影响计算
  3. 抑制型神经的编号排在兴奋型后
  4. 连接强度的值是配置文件(默认是test2.txt)的 Strength_CorEE 或 Strength_CorIE 或
     Strength_CorEI 或 Strength_CorII (相应于连接的两个神经元的类型) 与矩阵元素的乘积
  5. 多余的行不会被读入, 故可在末尾加任意注释.
  6. 神经元个数必须与连接矩阵的阶数相同, 否则...


代码内部的宏可调节的功能有:

默认神经元数目
#define Number_Exneuron 2
#define Number_Inneuron 0

神经网连接方式
#define CORTICAL_STRENGTH_NONHOMO 1

外部输入模式
#define POISSON_INPUT_USE 1

是否使用平滑上升电导
#define SMOOTH_CONDUCTANCE_USE 0

总计算时间(到时会停)
#define COMP_TIME  100000.0

解 ODE 的方法
#define RK4_RUN 1
#define RK3_RUN 0
#define RK2_RUN 0


为更改输出的信息, 以及一些重要的程序行为, 需要留意的代码如下(重要性高到低)

stdafx.h 全部, 特别是前半部. 用以设置如上各项宏

poisson_input.cpp
函数 void compute_perstep() 的后半部 及 函数 void LastRun().
用以更改 staffsave.txt, ras.txt, z.txt 等的输出内容

myopengl.cpp
void DrawSpikeInfo(). 更改屏幕额外显示的信息

datainput.cpp
int read_cortical_matrix(..). 更改读入临接矩阵的方式



changelog
更改日志

2012-05-22
  在非静默模式下(不带 -q 或 --quiet 命令行参数)的情况下使用全连矩阵(-mat -)时给出警告.
  输出文件的放电频率统计(--save-spike-interval FILE)改为使用15位精度输出.
  插值失败的检测改为实时进行----平均每秒不超过一次警告, 否则视为错误自动退出.
     (这是exponential I&F 计算时经常会出的问题)

2012-04-17
  以宏的方式添加了对 exponential I&F 模型的支持.

2012-02-25
  修正少许注释. 版本号更为 2.0.13

2011-12-02 (revision 17:18)
  用时间平均平滑掉 dt 带来的 spike 时刻电压的离散性. 仅在使用RC滤波时起作用(--RC-filter).

2011-11-23 (svn At revision 16)
  修正BUG:  检查文件写入冲突的部分可能会造成文件内容损坏.
  更正了一些注释.
  减少一些唠叨模式的输出文字.
  版本号增0.0.1

2011-10-23 ~ 2011-11-02
  进行了完整的命令行解析测试
  修正BUG:  -dt 参数修改无效
           --save-interval  不能设成与 -dt 相等
           文件输出前检查是否可用
  更改:  放电时刻文件 ras.txt 的神经元编号改为从 1 开始.
  更改行为: --pr-mul, --ps-mul, --psi-mul, 使用 a@b 之后的数默认是 c@(b+1)
  添加功能: 自定义 RC 低通滤波的系数: --RC-filter ci co
  添加功能(别名): -stv VALUE (同--save-interval VALUE)
  添加命令行参数 "-vv", 用以输出调试信息.

2011-10-03
发现 2011-09-06 的修改的 Bug: GUI 时参数的即时改变无效.
tempneu2 也改成一次生成的了.

2011-09-07
  解决 firing rate 超过 500次/s 的 Bug.
现在只剩下 spike_list 和 tempneu2 是动态生成的了.

2011-09-06
增加功能:
  可以用微秒极的系统时间作为随机种子(--seed-on), 并且这成为默认选项. (--verbose 时会显示其值)
  手动设置随机种子时, 允许用 0~2^31-1 的整数 或 0~1 之间的浮点数.
  可以通过文件读取每个神经的poisson输入率(相对值).
  可以从命令行直接读取相对poisson输入率和输入相度, 并且允许不完整指定. (--pr-mul, --ps-mul, --psi-mul)
优化:
  tempneu 变量预分配内存.

2011-09-05
对所有动态内存分配失败进行检测.
Xstream 系列 C++ 头文件限制在 datainput.cpp 和 myopengl.cpp 中使用.

2011-09-04
增加功能:
  命令行设置随机数种子.
  可以通过文件读取每个神经的poisson输入强度(相对值)(命令行参数给文件路径).

2011-08-10
修正保存数据的一个 bug: 现在可随意设定保存数据间隔.
增加 RC 滤波选项.

2011-07-11
修正命令行参数解析的一个 bug, 但未完全修正, 请完全按照指定格式使用.
添加对一些编译器的识别宏.
对程序在不同平台的运行效率进行了一个较完整的测试.

2011-05-29
基本把 poisson_input.cpp 的每个细节都弄清楚了.
去除了一些没用到的变量(数组).
把存放 Poisson input 的数组改成程序开头分配.
strobeupdate 改成只更新要用到的部分
一些代码改成更适于打印的形式
速度有约 10% 的提升

2011-05-20
添加参数 --save-spike-interval FILE

2011-05-11
添加参数 --save-spike FILE, --bin-save, --quiet

2011-05-06
总测试:
内存泄漏.
计算正确性.
异常的性能下降.

特例测试(时长=10000, 完全图)
编号  神经元数   输入强度  输入频率
1    1 + 0     弱        弱
2    1 + 0     弱        高
3    1 + 0     强        低
4    1 + 0     强        高

5    2 + 0     弱        弱
6    2 + 0     弱        高
7    2 + 0     强        低
8    2 + 0     强        高

9    20 + 0    中        中

1000 + 0  中        中    连接强度 中高


2011-05-05
在添加了一些命令行功能, 接口基本完整了. 取消原来想设的二进制格式输出.
把打开/关闭文件操作放在了开头/末尾, 效率问题大大改善.

2011-05-04
添加程序命令行参数功能(14 个, 部分未完成)
神经元个数已可参数输入

存在的问题
程序有过于频繁的内存申请, 释放. 导致效率显著下降.
把 tempneu, tmp_tempneu 的内存申请/分配提到开头和结尾,
速度提高约 21% (Linux, 29.2s -> 24.1s) 和 28% (WinXP, 78.6s -> 61.4s)
仍有一些变量有巨量的申请/释放.

2011-05-02, xyy
更改函数 void voltage_dt(..), 用宏区分是否使用 poisson 输入(明显提高计算速度).
去除 RYangle, RXangle 等变量
glColor3f(rcolor,gcolor,bcolor);
移出  glBegin(GL_QUADS);
删除函数
	glpolygon(..)  // 画多边形
	glring(..)  // 画环
	glarrow(..)  // 画箭头

变量名更改
GLUTWINDOWNUMBER  GLUT_WINDOW_ID
glut 回调函数的参数引用传递改为直接传递
按 Esc 的退出码改为 0 (原是 1)
貌似内存泄漏更严重了(18887040 bytes in 1 blocks are possibly lost).
但进程查看器中并未观察到内存使用的异常增长.
另外, 在纯计算模式时完全没有观察到内存泄漏.

2011-05-01, xyy
总计算时间改成可以从命令行读入(原来的宏成为默认值) (-t)
可以从命令行读入参数设置不显示图形界面. (-ng)
这两个功能的命令格式是:
raster_tuning -t TIME -ng
TIME 以毫秒为单位
在命令行窗口显示基本信息


2011-04-10, xyy

改成只有计算而没分析的程序.

改动:
文件名:
opengl.cpp, opengl.h -> myopengl.cpp, myopengl.h

更改:
vector 容量增长模式改为倍增(而不是线性增加, 见 void vector_reallocate(vector &a) @ loop.cpp, void raster_reallocate(raster &a) @ raster.cpp), 并增加简易的分配失败检查.
#define ENTER_KEY 13  // 而不是 10

去除的项(未能尽数列入)
FFTW_USE
FFT_TIME_LENGTH
MUL_LYAPUNOV_EVOLUTION
MUL_LYAPUNOV_COMP
MUL_EIG_TIME_INTERVAL
MUL_NUM_LYA
OPENGL_NOTUSE
#define SMALL_DISPLACE (1.0e-8)
#define FFTW_POWER_RECORD 1
#define CORTICAL_STRENGTH_ADJUST 1                                                      //   ------------1--->0
#define READ_DATA_CHECK 0
#define WRITE_DATA_CHECK 0
#define BIN_RECORD_USE 1
#define CORITCAL_INTERACTION_USE 1
LYA_STEP
#define MUL_LYAPUNOV_DEBUG 0
#define DEBUG_OUTPUT_ALL 0
#define VOT_DEBUG 0
#define FIRE_DEBUG 0
#define HEAD_NEXT_DEBUG 0
#define NO_SPIKE_DEBUG 0
#define ROOT_FIND_DEBUG 0
ROOT_FIND_DEBUG

void powerspectrum_vot_BINrecord();
void mul_lya_data_dump()
void mul_GramSchmit()
所有分析混沌的部分

去除所有未用到的变量

(It is safe to deallocate a NULL pointer)
if (a.vect_value != NULL) {
  free(a.vect_value);
}
改为 free()

// compute several lyapunov exponents **********************
long MUL_INI_SEED = 0;
struct neuron **mul_lya_neu = NULL;
struct raster *mul_lya_spike_list = NULL;
double ***mul_lya_perturb = NULL;
int MUL_VAR_NUM = 0;
int *mul_eig_num = NULL;
double *mul_eig_value = NULL;
double *mul_eig_value_square = NULL;
double *mul_norm_awaken_traject = NULL;
// this checks whether the reference and perturbed trajectories are at the same state
extern int num_equal_traject;

// compute several lyapunov exponents **********************
/* this part of program currently is designed for group neurons instead
of a single neuron, also it is for full lyapunov exponent in non-smooth
conductance case! */
// ramdom seed for generating random numbers
extern long MUL_INI_SEED;
// this saves the neurons's information for perturbed trajectories
extern struct neuron **mul_lya_neu;
// this saves the spike list for perturbed trajectories in each time step
extern struct raster *mul_lya_spike_list;
// this saves the perturbed vectors for computing Lyapunov exponents
extern double ***mul_lya_perturb;
/* number of perturbed variables, the order is: voltage, (exconductance),
(inconductance), (refractory residence time). */
extern int MUL_VAR_NUM;
// this saves the number of computing mulitiple local Lyapunov exponents
extern int *mul_eig_num;
// this saves the sum of multiple local Lyapuonv exponents
extern double *mul_eig_value;
// this saves the variance of multiple Lyapunov exponents
extern double *mul_eig_value_square;
// this saves the norm of perturbation for neurons outside refractory period
// after evolving the initial \epsilon different outside refractory period
extern double *mul_norm_awaken_traject;


周期记录
#if PERIOD_TRACE
VOT_TRACE
INPUT_RATE_ADJUST
/* whether or not traject the neuron's information(voltage,conductance,threshold)

stdafx.h
// stdafx.h : include file for standard system include files,
// or project specific include files that are used frequently, but
// are changed infrequently

#if !defined(AFX_STDAFX_H__A9DB83DB_A9FD_11D0_BFD1_444553540000__INCLUDED_)
#define AFX_STDAFX_H__A9DB83DB_A9FD_11D0_BFD1_444553540000__INCLUDED_

#define WIN32_LEAN_AND_MEAN       // Exclude rarely-used stuff from Windows headers
#define _CRT_SECURE_NO_DEPRECATE  // prevent MSVS say it's unsafe to use scanf etc

/// system headers
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <math.h>
#include <string.h>

#if defined(_WIN32) || defined(__WIN32__)
#  define _WINDOWS_USE_
#  include <windows.h>
#else  // Linux, no Mac OS support
#  undef _WINDOWS_USE_
#  include <unistd.h>
#  define Sleep(msec) usleep((msec)*1000)  // replace Sleep() in winbase.h (in Windows)
#endif

#include <GL/gl.h>
#include <GL/glu.h>
#include <GL/glut.h>

/// data structure
#include "raster.h"
#include "neuron.h"
#include "loop.h"

/// system wide constants
#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

//#ifndef INT_MAX
//#if __WORDSIZE != 16
//#define INT_MAX 2147483647
//#define UINT_MAX 4294967295U
//#else
//#define INT_MAX 32767
//#define INT_MAX 65535U
//#endif
//#endif

// these are used in open GL
#define ESCAPE_KEY 27
#define SPACE_KEY 32
#define ENTER_KEY 13
#define Q_KEY 113

// this is used to locate the drawn menu position when using open GL
#define menu_top 3

/// auxiliary macros
#define maximum(A,B) ((A) > (B) ? (A) : (B))
#define minimum(A,B) ((A) < (B) ? (A) : (B))
// usually B<C, keep B<=A<=C with up/down limite
#define crop(A,B,C) ((A) < (B) ? (B) : ((A) > (C) ? (C) : (A)))
// usually B<C, keep B<=A< C using periodization, i.e. A mod (C-B) + B, but only one period
#define periodize(A,B,C) ((A) < (B) ? (A) + (C) - (B) : ((A) >= (C) ? (A) - (C) + (B) : (A)))

#ifndef isnan
static inline int isnan(double x) { return x != x; }
#endif

// #ifdef DEBUG
#define P_NULL_ERR(p,st) {if ((p)==NULL) fprintf(stderr, "%s\n", (st));}
// might use perror(st) instead.
// variadic macros, introduced in C99
//#define P_ERR(...) {fprintf(stderr,__VA_ARGS__);}
//#define P_ERR(args...) {fprintf(stderr,##agrs);}
#define P_ERR(ps) {fprintf(stderr,ps);}

/// simulation settings and default constants
// whether use poisson input or current input
#define POISSON_INPUT_USE 1              ///XYY: modify this
// whether use the smoothed conductance (meaning that we use more than one ODE to
// describe the rising and decay time scale of conductance)
#define SMOOTH_CONDUCTANCE_USE 0         /// modify this
// whether consider the homogeneous cortical strength (a scalar independent of
// neuron index) or not! It is useful when using a N-by-N matrix to indicate the
// strength between neurons
#define CORTICAL_STRENGTH_NONHOMO 1      /// modify this

// check if using exponential I&F model!
#define EXPONENTIAL_IF_USE 1
#if EXPONENTIAL_IF_USE
#  define EIF_CUBIC_INTERPOLATION 1
#else
#  define EIF_CUBIC_INTERPOLATION 0
#endif

// maximum number of neurons that can accepted
#define NUM_NEU_MAX (32767)
// the number of excitatory neurons
#define Number_Exneuron 2                ///XYY: modify this
// the number of inhibitory neurons
#define Number_Inneuron 0
// the final stopping time of computation in millisecond
#define COMP_TIME  10000.0               /// modify this
// saving time length (the interval is [T-20, T]) for voltage, conductance
// and also new raster events
//#define TIME_LENGTH 20.0

// decay time scale of Ex_conductance
#define Time_ExCon      2.0
// if we use smoothed excitatory conductance instead of jump conductance, we
// should have another ODE. this is the time scale in that ODE
#define Time_ExConR     0.5
// decay time scale of In_conductance
#define Time_InCon      5.0
// the same thing but for inhibitory conductance
#define Time_InConR     0.8
// the length of refractory period
#define TIME_REFRACTORY 2.0

#if EXPONENTIAL_IF_USE
#define VOT_RESET  (-0.375)
#define Vot_Threshold  (4.375)
#define VOT_TAKEOFF (0.625)
#define VOT_DELTAT (0.4375)
#else
// the reset value for voltage after spike
#define VOT_RESET       0.0
// the threshold value for voltage
#define Vot_Threshold   1.0
#endif

// the resting value of leakage conductance
#define Vot_Leakage     0.0
// the resting value of excitatory conductance
#define Vot_Excitatory  (14.0/3.0)
// the resting value of inhibitory conductance
#define Vot_Inhibitory  (-2.0/3.0)
// the leakage conductance "G^L"
#define Con_Leakage     0.05

#if SMOOTH_CONDUCTANCE_USE
// this means using two ODEs to describe the dynamics of conductance
// therefore, the conductance is smooth with both rising and decay time scale
#define Stepsmooth_Con 2
#else
// this means using only one ODE to describe the dynamics of condcuctance
// therefore, the conductance only has decay time scale and infinitely fast
// rising time scale (instantaneous jump)
#define Stepsmooth_Con 1
#endif

// the length of "data" stored in the neuron structure
// the first is saving voltage, the last is saving refractory residence time
// the middle saves the conductance part, there are maybe three cases:
// (1) Ex_conductance after two step of smoothness, (2) Ex_conductance after one
//     step of smoothness, (3) Ex_conductance without smoothness
// (1) In_conductance after two step of smoothness, (2) In_conductance after one
//     step of smoothness, (3) In_conductance without smoothness
#define size_neuronvar  (2*Stepsmooth_Con+2)

////////////////////////////////////////////////////////////
// simulator behaviour variables

extern int RUN_DONE;           // whether the running program is finished or not
extern bool g_no_graphic;      // whether use graphic mode
extern bool g_b_verbose_debug;
extern bool g_b_quiet;
extern bool g_b_save_while_cal;
extern bool g_b_save_use_binary;
extern bool g_b_RC_filter;     // use RC filter or not
extern double g_RC_filter_ci;  // filter coefficient:
extern double g_RC_filter_co;  // output[t] = co*output[t-dt]+ci*input[t]

extern char g_ras_path[];
extern char g_spike_interval_path[];
extern FILE *g_fout;
extern FILE *g_cond_out;

////////////////////////////////////////////////////////////
// Global variables for GUI

// these are global variables used in open GL
extern GLfloat xdepth;
extern GLfloat ydepth;
extern GLfloat zdepth;
extern int GLUT_WINDOW_ID;
extern int GLOBAL_WINDOW_WIDTH;
extern int GLOBAL_WINDOW_HEIGHT;
extern int FULL_SCREEN;

extern int GRAYSCALE;
// the initial time for the run
extern double GLOBAL_TI;
// this is the whole time length of data shown in the window screen
// normally, this is choosen as 512 ms.
extern int WINDOW_BIN_LENGTH;
// this number indicates the position of cursor in the menu
extern int FIDDLE_PARAMETER;
// this indicates after how many time steps, for the data to show on the screen
extern int STEPS_PER_DRAW;
extern int STD_VIEW;
// this is the strobe structure used in open GL
extern struct strobe **GLOBAL_STRA;
// this number indicates the corresponding variable shown in the window screen
extern int DRAW_FLAG;

////////////////////////////////////////////////////////////
// simulation related global variables

// this is the time length of each data point to show, usually, this
// is chosen as 0.5 or 1 ms
extern double SLIGHT_BIN;
// The following variables are the input reading from input file!
// this is time step used in the run
extern double Tstep;
// i.e. COMP_TIME
extern double g_comp_time;
// this is the rate either in the current input case or poisson input case
// this saves the current time of computation, the dimension is "ms"
extern double time_evolution;
// this saves the starting time of the run, normally it is chosen as zero
extern double last_time;

////////////////////////////////////////////////////////////
// simulation global data variables

// this saves the spike list in each time step
extern struct raster spike_list;
// this saves all the spike event during the computation
extern struct raster RAS;
// this saves neuronal network information
extern struct neuron *neu;
// this saves the generated Poisson spike train for each neuron
extern struct vector *poisson_input;
// this saves the first spike time of Poisson input outside the current time step
extern double *last_input;

////////////////////////////////////////////////////////////
// neuron related global variables

extern int g_num_neu;
extern int g_num_neu_ex;
extern int g_num_neu_in;
extern double Rate_input;
#if POISSON_INPUT_USE
// this is the Poisson input strength to excitatory conductance
// it is not used in current input case
extern double Strength_Exinput;
// the same as above but the Poisson input strength to inhibitory conductance
// normally, it is chosen as zero
extern double Strength_Ininput;
#else
// in current input case, the sinusoidal drive is given by
// Current_0 + Current_1*sin(2*pi*Rate_input*t+phase)
extern double Current_0;
extern double Current_1;
extern double *phase;
#endif

#if CORTICAL_STRENGTH_NONHOMO
// this matrix saves the cortical strength between neurons in the network
extern double** cortical_matrix;
#endif

// the strength of excitatory neuron to excitatory neuron
extern double Strength_CorEE;
// the strength of excitatory neuron to inhibitory neuron
extern double Strength_CorIE;
// the strength of inhibitory neuron to inhibitory neuron
extern double Strength_CorII;
// the strength of inhibitory neuron to excitatory neuron
extern double Strength_CorEI;

////////////////////////////////////////////////////////////
// temporarily space in computation

// random seed for generating Poisson process
extern long *initialseed_neuron;
// these are used for generating random variables
extern long initial_pertub_Vot;
extern long initial_pertub_Ex;
extern long initial_pertub_In;
#if SMOOTH_CONDUCTANCE_USE
extern long initial_pertub_Ex_H;
extern long initial_pertub_In_H;
#endif
extern unsigned int initial_seed;
extern long* ran_iy;
extern long** ran_iv;

#if POISSON_INPUT_USE
extern double *g_arr_poisson_rate;
extern double *g_arr_poisson_strength_E;
extern double *g_arr_poisson_strength_I;
extern int *g_begin_poisson_index;
extern int *g_tempbegin_poisson_index;
extern int *tmp_tempbegin_poisson_index;
#endif
extern neuron *tmp_tempneu;
extern neuron *tmp_tempneu2;

// macro defintion for neurons with excitatory type and inhibitory type
#define Type_Exneuron  1
#define Type_Inneuron  0
// macro definition for neurons' state, whether it is ouside refractory or inside it
#define STATE_ACTIVE  1
#define STATE_REFRACTORY  0

// estimated maximum times for searching a root, it is used in root_search function
#define Maxnum_search  50
// estimated maximum numerber of poisson input in each time step, if the input rate
// is really high, we should choose much larger number!
#define Maxnum_input  200
// the tolerance error for root searching
#define root_acc  (1.0e-12)
// the tolerance error to differ two values
#define LOW_BOUND (1.0e-16)

// this number means the least tolerance space in the raster structure
// if the space left in the raster structure is less than this number, the new space
// will be assigned to the raster structure
#define RASTER_ENDURE_SIZE  5
// additional assigned space for raster structure
#define RASTER_ADD_SIZE 50
// additional assigned space for vector structure
#define VECTOR_ADD_SIZE 50
// initial estimate data length for raster structure (for saving spike events in
// each time step)
#define RASTER_SPIKE_SIZE 20

// this uses different orders of Runge-Kutta algorithm
#define runge_kutta runge_kutta4

// macro definition for random number generating functions
#define RANDOM ran1
#define NTAB 32 // ran1() using

#endif // !defined(AFX_STDAFX_H__A9DB83DB_A9FD_11D0_BFD1_444553540000__INCLUDED_)
